### ISSUE TITLE
Fusion Kernel Accuracy Issues

### ISSUE THREAD

**Issue by jeromeku** [pytorch/pytorch#122040]
### 🐛 Describe the bug

Getting accuracy issues when trying to fuse elementwise optimizer ops such as for `AdamW`.

Below is a minimal script comparing a compiled vs reference torch version:

```python
import torch

M, N = 4096, 4096
dtype = torch.float32
BETA1, BETA2 = 0.9, 0.999
EPS = 1e-8
torch.manual_seed(0)


def make_data(M, N, dtype):
    exp_avg = torch.randn(M, N, device="cuda", dtype=dtype)
    exp_avg2 = exp_avg**2
    grad = torch.randn(M, N, device="cuda", dtype=dtype)
    return exp_avg, exp_avg2, grad


def test_fn(exp_avg, exp_avg2, grad, beta1=BETA1, beta2=BETA2, eps=EPS):
    exp_avg.mul_(beta1).add_(grad, alpha=(1.0 - beta1))
    exp_avg2.mul_(beta2).addcmul_(grad, grad, value=1.0 - beta2)
    denom = exp_avg2.sqrt().add_(eps)
    norm_grad = exp_avg / denom
    return exp_avg, exp_avg2, denom, norm_grad


def run_reference(
    exp_avg,
    exp_avg2,
    grad,
    beta1=BETA1,
    beta2=BETA2,
    eps=EPS,
):

    exp_avg, exp_avg2, denom, grad = test_fn(
        exp_avg, exp_avg2, grad, beta1=beta1, beta2=beta2, eps=eps
    )

    return exp_avg, exp_avg2, denom, grad


def run_inductor(
    exp_avg,
    exp_avg2,
    grad,
    beta1=BETA1,
    beta2=BETA2,
    eps=EPS,
    fullgraph=False,
    dynamic=False,
    mode="default",
):
    compiled = torch.compile(test_fn, fullgraph=fullgraph, mode=mode, dynamic=dynamic)

    exp_avg, exp_avg2, denom, grad = compiled(
        exp_avg, exp_avg2, grad, beta1=beta1, beta2=beta2, eps=eps
    )

    return exp_avg, exp_avg2, denom, grad


exp_avg, exp_avg2, grad = make_data(M, N, dtype)

ref_out = run_reference(exp_avg, exp_avg2, grad)
inductor_out = run_inductor(exp_avg, exp_avg2, grad)

for ref, ind in zip(ref_out, inductor_out):
    print(f"max delta: {torch.max(torch.abs(ref - ind))}")
```

This gives the following outputs:
``` 
max delta: 0.0
max delta: 0.0
max delta: 0.06131890416145325
max delta: 1.403359055519104
```

Changing the `dtype` to `torch.half` gives the following:
```
max delta: 0.0
max delta: 0.0
max delta: 0.061279296875
max delta: inf
```

### Versions

Collecting environment information...
PyTorch version: 2.3.0.dev20240312+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 17.0.0 (++20230807073112+34cf263e6149-1~exp1~20230807073225.14)
CMake version: version 3.27.2
Libc version: glibc-2.35

Python version: 3.11.4 (main, Jun  7 2023, 12:45:48) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-6.1.0-060100-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3050 Laptop GPU
Nvidia driver version: 545.29.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.5
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn.so.8.2.4
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.4
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.4
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.4
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.4
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.4
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       GenuineIntel
Model name:                      11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
CPU family:                      6
Model:                           141
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        1
CPU max MHz:                     4600.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4608.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       384 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        10 MiB (8 instances)
L3 cache:                        24 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] lion-pytorch==0.0.8
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] pytorch-triton==3.0.0+a9bc1a3647
[pip3] torch==2.3.0.dev20240312+cu121
[pip3] triton==3.0.0
[conda] No relevant packages

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang
